docs/interpreter_checklist.md

```diff
@@ -0,0 +1,67 @@
+   the single source of truth, and help to track gaps in those implementations
+   back to the spec.
+
+   **Why Unique**: The fact that the spec will be augmented with Dynamism,
+   Quantization can affect the suffix number in the constraint label or might
+   change it to something else. A unique identifier will enable faster updates
+   to comments we are planning to add here.
+   1. It is OK to have a comment with multiple constraint-labels or to have
+      multiple comments with the same constraint-label. It all depends on how
+      the constraint-label(s) are coded in the implementations.
+   1. In case there is a mismatch between the constraints in the
+      implementation VS those in the specification, make sure there is an
+      open issue reflecting that discrepancy.
+1. Make sure that the there are sufficient tests to cover for the
+   implementations (e.g. verifiers, type inference, interpreter) of the
+   specification. Specifically,
+   1. Regarding interpreter tests
+      1. For interpreter: Write tests following the
+         [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/reference.md#testing-guidelines).
+      1. Add a link of the test file in the spec (e.g.
+         [More Examples](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#add)).
+   1. For verifier and type inference: Affecting [this file](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/ops_stablehlo.mlir).
+      1. Make sure that there is at least one test for each constraint in
+         verifier and type inference methods. These tests will mostly be
+         negative tests exercising the fact that the constraints are not met.
+         Make sure to add at least one positive tests in case there are none.
+         Note that it is out of scope to revisit the correctness of type
+         inference as in
+         [file](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/infer_stablehlo.mlir).
+      1. Make sure that all the tests related to the op under test, are placed
+         together.
+      1. Make sure that all the tests related to the op under test, are
+         prepended with `CHECK-LABEL` lit macro.
+      1. Chose the function name of the tests using the format
+         `op_name_Cn1_Cn2_...` for a function testing constraints `Cn1`, `Cn2`
+         for op `op_name`. In cases when the proposed format does not apply,
+         keep the existing name.
+1. Locally run the
+   [ccov](https://github.com/openxla/stablehlo/blob/main/build_tools/github_actions/ci_build_stablehlo_code_coverage.sh)
+   tool to make sure that the followings yield >= 90% coverage. If not, then add
+   tests to achieve the goal.
+   1. Code and tests introduced for implementing the interpreter.
+   1. The existing code and tests for verifier and type inference.
+1. Once all the items are finsihed, update the interpreter column in status.md
+   to `yes`.
```

Comment:
This is a good checklist, but I think it can be sharpened up to make instructions more straightforward, similarly to the spec checklist. Here's how I would've written it:

```
While implementing the interpreter:

1. Consult [hlo_evaluator](link) and [mlir_interpreter](link) to identify tricky
   implementation details and potential functionality gaps.

2. File tickets for the corresponding software components if you find any bugs
   or missing functionality.

After implementing the interpreter:

1. In [StablehloOps.td](link):
  * Make sure that the `summary` in op's ODS follows the standard format.

2. In [TypeInference.cpp](link):
  * Delete comments that say things like "Verify the following properties: ...".
  * Add comments that reference constraint labels from the spec in the format
    `<op_name>_Cn`, e.g. `SliceOp_C1` to identify which parts of verifiers and 
    shape functions correspond to which constraints in the specification.
      * It is OK to have a comment with multiple constraint labels or to have 
        multiple comments with the same constraint label. It all depends on how 
        the constraints are implemented.
      * In case there is a mismatch between the constraints in the 
        implementation VS those in the specification, make sure there is an open 
        issue reflecting that discrepancy.

3. In [interpreter tests](link to the directory):
  * Add file called `interpret_<op_mnemonic>.mlir`.
  * Write tests following the [testing guidelines](link).

4. In [ops_stablehlo.mlir](link):
  * Make sure that there is at least one test for each constraint...
  * Make sure that all the tests related to the op under test...
  * Make sure that all the tests related to the op under test...
  * Chose the function name of the tests using the format...
  * Keep adding tests until the [ccov](link) shows >= 90% coverage for the op.

5. In [spec.md](link):
  * Add a link to `interpret_<op_mnemonic>.mlir` to the "Examples" section.

6. In [status.md](link):
  * Update the "Interpreter" column to `yes`.
============

Reply:
4. Should we need to add test for constraints covered by the ODS? What do we do with tests covered in Types section (e.g. 1-dimensional tensor... and the test "x should be rank 1...") Should we add them (if we don't have it and not covered by ODS)?

Reply:
That is right.  we should add tests only if if we don't have it and not covered by ODS. IMO, It does not worth testing semantics of various mlir traits.

Reply:
Agreed.